Add option to return an exit-code when occ status signals an update is needed

leegarrett opened the pull request:
Signed-off-by: Lee Garrett <lgarrett@rocketjump.eu>

* Resolves: #35704

## Summary
This is an improved version of my earlier PR  #35830. This is now a parameter to `status`, so it allows for adding more return codes later, which can then for example be consumed in systemd units, or monitoring checks.

It adds a parameter that can be used in scripts. `./occ status -e` returns 0 when operating normally, and 1 if there is maintenance mode, 2 if `./occ upgrade` is required.

## TODO

- [ ] Test coverage?

## Checklist

- [x] Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [x] [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

leegarrett committed:
Implement occ status command via return codes (Fixes: #35704)

Running `./occ status -e` will produce any output. However, it will:

exit 0 during normal operation,
exit 1 when in maintenance mode,
exit 2 when `./occ upgrade` is needed.

Signed-off-by: Lee Garrett <lgarrett@rocketjump.eu>

nickvergessen commented on core/Command/Status.php:
```diff
@@ -47,22 +48,46 @@ protected function configure() {
+			# everything operating normally
+				return 0;
+			}
+		} else {
```
```suggestion
		}

```

nickvergessen commented on core/Command/Status.php:
```diff
@@ -47,22 +48,46 @@ protected function configure() {
 		];
 
 		$this->writeArrayInOutputFormat($input, $output, $values);
+		}
```
```suggestion
```

leegarrett replied:
Both fixed in the newest branch push.

nickvergessen reviewed (commented):
Otherwise you would need to add a tab on each line

come-nc commented on core/Command/Status.php:
```diff
@@ -47,17 +48,40 @@ protected function configure() {
+
+		if ($input->getOption('exit-code')) {
+
+			if ($maintenanceMode === true) {
+			# checks maintenance mode first
+				return 1;
+
```
```suggestion
			if ($maintenanceMode === true) {
				/* Checks maintenance mode first */
				return 1;

```
And same change for the 2 other comments.

leegarrett replied:
I've push a new commit that fixes the indentation of all three comments.

come-nc commented:
@leegarrett Code style checker still complains, this time about empty lines and spacing: https://github.com/nextcloud/server/actions/runs/3824719618/jobs/6507559027

Also for the comments I’d prefer if you avoid using `#`.

come-nc commented:
Also last commit is missing signoff and DCO complains: https://github.com/nextcloud/server/pull/35873/checks?check_run_id=10400685089

leegarrett commented:
> @leegarrett Code style checker still complains, this time about empty lines and spacing: https://github.com/nextcloud/server/actions/runs/3824719618/jobs/6507559027
> 
> Also for the comments I’d prefer if you avoid using `#`.

I've hopefully fixed the issue now. Since the 2nd commit wasn't signed off, I've now just squashed it with the first, removed the extra whitespace, and removed the comments. They didn't add much information, anyway.

come-nc commented on core/Command/Status.php:
```diff
@@ -47,17 +48,33 @@ protected function configure() {
 	}
 
 	protected function execute(InputInterface $input, OutputInterface $output): int {
+		$maintenanceMode = $this->config->getSystemValueBool('maintenance', false);
+		$needUpgrade = Util::needUpgrade();
+		if ($input->getOption('exit-code')) {
+			if ($maintenanceMode === true) {
+				return 1;
+			} elseif ($needUpgrade == true) {
+				return 2;
+			} else {
+				return 0;
+			}
+		}
```
Would it make sense to move this to after the call to writeArrayInOutputFormat ?
Or is there a reason to remove output when having the exit code?

nickvergessen replied:
Done

leegarrett replied:
Yes, you don't want output e.g. in `ExecCondition` for a system service, as this will also be logged in the unit every time and pollutes the logs unnecessarily. We can however add it back conditionally with `-v` if you see any use of this output.

come-nc reviewed (approved):
See my question on output, but otherwise looks good, codesniffer and DCO are both happy. Failure in drone is unrelated.

nickvergessen committed:
Print output also with status code

Signed-off-by: Joas Schilling <coding@schilljs.com>

leegarrett commented:
Not sure if resolved conversations get notified, so here's a copy/paste of it:

Yes, you don't want output e.g. in `ExecCondition` for a system service, as this will also be logged in the unit every time and pollutes the logs unnecessarily. We can however add it back conditionally with `-v` if you see any use of this output.

nickvergessen commented:
> Not sure if resolved conversations get notified, so here's a copy/paste of it:

They do,

> Yes, you don't want output e.g. in ExecCondition for a system service, as this will also be logged in the unit every time and pollutes the logs unnecessarily. We can however add it back conditionally with -v if you see any use of this output.

That makes sense. Will make it so.

nickvergessen committed:
No output by default when --exit-code is used

Signed-off-by: Joas Schilling <coding@schilljs.com>

leegarrett commented:
Hello, can we also get this backported to v25?

nickvergessen commented:
/backport to stable25

solracsf commented:
No documentation on this?

nickvergessen commented:
Should be added to https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/occ_command.html yes.
@leegarrett  could you extend the documentation at https://github.com/nextcloud/documentation/blob/master/admin_manual/configuration_server/occ_command.rst
Otherwise raising an issue in https://github.com/nextcloud/documentation/issues with a link to this PR would be helpful too.

leegarrett commented:
FYI, the doc update PR is at https://github.com/nextcloud/documentation/pull/9583

DaphneMuller commented:
hello @leegarrett 
Thank you so much for your work on this pull request! This ticket seems to still have the tag 'pending documentation'. 
Is there any chance you could clarify what documentation is missing? Is this for admins or for app developers?

nickvergessen commented:
It's linked above https://github.com/nextcloud/documentation/pull/9583
